### ISSUE TITLE
Not immune to while(1){}

### ISSUE THREAD

**Issue by XmiliaH** [patriksimek/vm2#180]
Following code will not time out.
```JS
const {VM} = require('vm2');
new VM({timeout:1}).run(`
		function main(){
		while(1){}
	}
	new Proxy({}, {
		getPrototypeOf(t){
			global.main();
		}
	})`);
```

**Comment by anonyco** [comment]
There is likely no good solution to this. NodeJS is single process. while(1){} protection requires multi-process so that the checker does not get blocked by the while(1){}. Maybe if we could rekindle the SpiderNode project.......

**Comment by iamahuman** [comment]
If spawning a separate process is not desired, you may find https://nodejs.org/api/worker_threads.html useful.

**Comment by iamahuman** [comment]
This is caused by `Decontextify.value` testing for the guest object's prototype with `instanceOf`. Node.js `vm` module is of course immune to this. A poor man's patch might be to backdoor `Proxy` so that we could detect any Proxy objects from the sandbox, but an ideal approach would be to just never trust any objects and functions from it and put contextify.js into sandbox, using primitive values to communicate and handles for referencing guest objects.

**Comment by joelgriffith** [comment]
Just for folks to see, browserless uses a child-process and message-passing to talk with the child running the untrusted code. It's a bit of a setup task, but once done you'll be immune to `while(1){}` attacks. The [source is here](https://github.com/joelgriffith/browserless/blob/master/src/Sandbox/index.ts).

**Comment by iamahuman** [comment]
The whole point of this project seems to be to not spawn a whole separate
OS process. Otherwise there are many solutions available which are much
more secure.

**Comment by joelgriffith** [comment]
> Otherwise there are many solutions available which are much more secure.

Which ones? I'm not sure there's a better alternative to VM2 in NodeJS, at least not that I've seen

**Comment by chung-leong** [comment]
The maintainer of this library pointed to this alternative in issue #80:

https://github.com/laverdet/isolated-vm

I haven't tried it yet myself.

**Comment by tian-ma** [comment]
I only run untrusted code in child process. Can't risk running it in the main V8 instance especially when you know it's single threaded.

The problem of calling V8 engine directly is lack of importing module. In my case I need to import request module to allow HTTP operations. It also has to be asynchronous.

I find vm2 + child_process a good combination.

**Comment by idan-at** [comment]
Hey,

myself and @harelmo have created a library on top of vm2 that is immune to `while(1) {}`.

Similar to what tian-ma mentioned, but we use node.js worker threads instead of child processes.

take a look here: https://www.npmjs.com/package/isolated-runtime and let us know what you think :)

**Comment by Yogu** [comment]
Another workaround would be to just disable `Proxy` within the sandbox, right?

    const vm = new VM({
        sandbox: {
            Proxy: undefined
        },
        timeout: 100
    });

**Comment by XmiliaH** [comment]
Using Promise it is also possible to write code that never times out:
```JS
"use strict";
const {VM} = require('vm2');
const untrusted = '(' + function(){
	Promise.resolve().then(a=>{
		while(1){}
	});
}+')()';
try{
	console.log(new VM().run(untrusted));
}catch(x){
	console.log(x);
}
```
Problem here is that .then will queue the function to be called later in the microtask queue.
To fix this we need to write our own Promise.then & related functions or disable Promise.

**Comment by XmiliaH** [comment]
Also this works:
```JS
const {VM} = require('vm2');

const script = '(' + function() {
	(async function x(){
		await {then: y=>y()};
		while(1){}
	})();
}+')()';

new VM({timeout:10}).run(script);
```
So just overriding Promise won't work.
No idea how to fix this.

**Comment by mmiszy** [comment]
While this remains open, we should add a clarification to the README to warn potential users. This vulnerability allows one to escape the sandbox and thus it defeats the purpose of the library.

**Comment by XmiliaH** [comment]
Readme was updated https://github.com/patriksimek/vm2/commit/77f5265ab53b87864a312156ee62b1082787e9b0#diff-04c6e90faac2675aa89e2176d2eec7d8. And I don't know how this could be used to escape the sandbox, you only can escape the intended timeframe the script should run in.

**Comment by crowder** [comment]
Perhaps this should be its own issue, but why does NodeVM not support "timeout"?

**Comment by XmiliaH** [comment]
@crowder since in NodeVM the setTimeout, setInterval & setImmediate exists, it would be easy to circumvent the timeout. Maybe there are other reasons too, but I don't work with NodeVM, only VM.

**Comment by szydan** [comment]
@XmiliaH  wouldn't implementing a timeout prevent at least from unintended infinite loops? 
The malicious while(true) + setTimeout combination would still be an issue but any simple unintended infinite loop done by mistake could be handled by supporting timeout

**Comment by XmiliaH** [comment]
My concern is that NodeVM allowes to require modules which might be host modules and timing out while they change global state might be bad. Also the NodeVM returns a module and users quite likely would call functions exposed by the module, however the timeout would only apply to the module loading and not the function calls. 
If you want to call a function with a timeout you can just implement it yourself like
```JS
function doWithTimeout(fn, timeout) {
	let ctx = CACHE.timeoutContext;
	let script = CACHE.timeoutScript;
	if (!ctx) {
		CACHE.timeoutContext = ctx = vm.createContext();
		CACHE.timeoutScript = script = new vm.Script('fn()', {
			filename: 'timeout_bridge.js',
			displayErrors: false
		});
	}
	ctx.fn = fn;
	try {
		return script.runInContext(ctx, {
			displayErrors: false,
			timeout
		});
	} finally {
		ctx.fn = null;
	}
}
```

**Comment by szydan** [comment]
Good point @XmiliaH and thanks for the example 
I see your example is using vm package and not vm2 
Would be nice to be able to use vm2 and still be able to timeout it out

**Comment by XmiliaH** [comment]
@szydan The code I posted is used by VM2 to timeout the VM. If the NodeVM had a timeout too, it would use the same functionality.

**Comment by dejavughost** [comment]
> Readme was updated [77f5265#diff-04c6e90faac2675aa89e2176d2eec7d8](https://github.com/patriksimek/vm2/commit/77f5265ab53b87864a312156ee62b1082787e9b0#diff-04c6e90faac2675aa89e2176d2eec7d8). And I don't know how this could be used to escape the sandbox, you only can escape the intended timeframe the script should run in.

@XmiliaH I am a bit confused now. Does this limitation still apply?

Regarding the following statements. Could you please elaborate a bit more, so that I can understand?
> `timeout` - Script timeout in milliseconds. **WARNING:** You might want to use this option together with `allowAsync=false`

> IMPORTANT: Timeout is only effective on synchronous code that you run through run. Timeout does NOT work on any method returned by VM.

**Comment by dejavughost** [comment]
```js
  const vm = new VM({
    timeout: 500, // specifies the number of milliseconds to execute code before terminating execution
    allowAsync: true
    // [...]
  })
```

Does this permutation (`allowAsync=true` and `timeout=500`) work as expected in all known contexts for `async` code? For example:

```js
(async function () {
  const result = await db.getData()
})()
```

**Comment by XmiliaH** [comment]
With `allowAsync=true` and a timeout the following code
```js
(async function () {
  while(1){}
})()
```
would run the main chunk with the timeout but all async micro tasks are not affected by the timeout, so the while loop would not be stopped.

**Comment by arleytm** [comment]
> so the while loop would not be stopped.

[XmiliaH](https://github.com/XmiliaH) In the following scenario, the script execution timed out after 500ms 🤷‍♂️. At least this is the message thrown by the VM.

```js
const { VM } = require('vm2')

async function run(code) {
  const vm = new VM({
    timeout: 500, // specifies the number of milliseconds to execute code before terminating execution
    allowAsync: true
  })

  try {
    await vm.run(code)
  } catch (error) {
    console.log(error.message)
  }
}

const code = `(async function () {
  while (true) {}
})()`

run(code)

```

**Comment by XmiliaH** [comment]
Try with:
```js
const code = `(async()=>{
  await(async()=>{})();
  while(true){}
})()`
```

**Comment by arleytm** [comment]
> ```js
> const code = `(async()=>{
>   await(async()=>{})();
>   while(true){}
> })()`
> ```

@XmiliaH Thanks for the clarification 👍.

I debugged the `vm2.run(...)` method and found that it uses the [`script.runInContext(contextifiedObject[, options])`](https://nodejs.org/api/vm.html#scriptrunincontextcontextifiedobject-options) API under the hood. Interestingly, currently, in the documentation of the vm (Node.js core module), the async/timeout limitation seems not to be mentioned.

https://github.com/patriksimek/vm2/blob/62062adbd7145f734615c9bce190b739ecb86059/lib/vm.js#L281-L290

In this scenario (in Node.js core vm land), I would have expected the execution to be terminated, and an [Error](https://nodejs.org/api/errors.html#class-error) to be thrown and then be propagated to the vm2 bridge. I have reproduced the same behavior using the node core vm module. For example:

```js
const vm = require('node:vm')
const context = vm.createContext()
const code = `
(async () => {
  await (async () => {})()
  while (true) {}
})()
`
const scriptCode = new vm.Script(code)
scriptCode.runInContext(context, { timeout: 500 }) // this run does not timeout after 500 ms
```

Do you know if the Node.js runtime team is aware of this limitation?

**Comment by XmiliaH** [comment]
They are aware (see https://nodejs.org/api/vm.html#timeout-interactions-with-asynchronous-tasks-and-promises).

**Comment by XmiliaH** [comment]
I know of the option. However, I did not really look into it as it is not bulletproof and we use a passthrough context for the timeout which should only capture the microtasks from that context and not the sandbox one so it would need a redesign of the timeout functionality.
https://github.com/patriksimek/vm2/blob/62062adbd7145f734615c9bce190b739ecb86059/lib/vm.js#L121-L140
